File: Voter Registration/01_scraper.py
import camelot
import os
import glob
import pandas as pd
import csv

pdf_dir = os.path.join(os.path.dirname(__file__), "data_pdf")
csv_dir = os.path.join(os.path.dirname(__file__), "data_csv")
os.makedirs(csv_dir, exist_ok=True)

for pdf_file in glob.glob(os.path.join(pdf_dir, "*.pdf")):
    base_filename = os.path.splitext(os.path.basename(pdf_file))[0]
    csv_path = os.path.join(csv_dir, f"{base_filename}.csv")

    if os.path.exists(csv_path):
        print(f"Already extracted, skipping: {base_filename}.pdf")
        continue

    tables = camelot.read_pdf(pdf_file, pages="all")

    if tables:
        combined_rows = []

        for i, table in enumerate(tables):
            rows = table.df.values.tolist()

            # Skip the first row in subsequent tables (duplicate header)
            if i > 0:
                rows = rows[1:]

            combined_rows.extend(rows)

        with open(csv_path, "w", newline="") as f:
            writer = csv.writer(f)
            writer.writerows(combined_rows)
        print(f"Written: {base_filename}.csv")
    else:
        print(f"No tables found: {base_filename}.pdf")

File: Voter Registration/01_scraper.R
library(tabulapdf)

pdf_dir <- "Voter Registration/data_pdf"
csv_dir <- "Voter Registration/data_csv"

pdf_files <- list.files(pdf_dir, pattern = "\\.pdf$", full.names = TRUE)

for (pdf_file in pdf_files) {
  base_name <- tools::file_path_sans_ext(basename(pdf_file))
  csv_path  <- file.path(csv_dir, paste0(base_name, ".csv"))

  if (file.exists(csv_path)) {
    message("Already extracted, skipping: ", basename(pdf_file))
    next
  }

  tables <- extract_tables(pdf_file, output = "matrix")

  if (length(tables) == 0) {
    message("No tables found: ", basename(pdf_file))
    next
  }

  # First table: keep all rows (includes header)
  # Subsequent tables: drop first row (duplicate header)
  combined <- do.call(rbind, c(
    list(tables[[1]]),
    lapply(tables[-1], function(t) t[-1, , drop = FALSE])
  ))

  write.csv(combined, csv_path, row.names = FALSE)
  message("Written: ", basename(csv_path))
}
